pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/BookkeeperBucketSnapshotStorage.java

```diff
@@ -0,0 +1,247 @@
+    public void start() throws Exception {
+        pulsar.getBookKeeperClient();
+        this.bookKeeper = pulsar.getBookKeeperClientFactory().create(
+                pulsar.getConfiguration(),
+                pulsar.getLocalMetadataStore(),
+                pulsar.getIoEventLoopGroup(),
+                Optional.empty(),
+                null
+        );
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (bookKeeper != null) {
+            bookKeeper.close();
+        }
+    }
+
+    private CompletableFuture<Void> addSnapshotSegments(LedgerHandle ledgerHandle,
+                                                        List<SnapshotSegment> bucketSnapshotSegments) {
+        List<CompletableFuture<Void>> addFutures = new ArrayList<>();
+        for (SnapshotSegment bucketSnapshotSegment : bucketSnapshotSegments) {
+            addFutures.add(addEntry(ledgerHandle, bucketSnapshotSegment.toByteArray()));
+        }
+
+        return FutureUtil.waitForAll(addFutures);
+    }
+
+    private CompletableFuture<SnapshotMetadata> parseSnapshotMetadataEntry(LedgerEntry ledgerEntry) {
+        CompletableFuture<SnapshotMetadata> result = new CompletableFuture<>();
+        try {
+            result.complete(SnapshotMetadata.parseFrom(ledgerEntry.getEntry()));
+        } catch (IOException e) {
+            result.completeExceptionally(e);
+        }
+        return result;
+    }
+
+    private CompletableFuture<List<SnapshotSegment>> parseSnapshotSegmentEntries(
+            Enumeration<LedgerEntry> entryEnumeration) {
+        CompletableFuture<List<SnapshotSegment>> result = new CompletableFuture<>();
+        List<SnapshotSegment> snapshotMetadataList = new ArrayList<>();
+        try {
+            while (entryEnumeration.hasMoreElements()) {
+                LedgerEntry ledgerEntry = entryEnumeration.nextElement();
+                snapshotMetadataList.add(SnapshotSegment.parseFrom(ledgerEntry.getEntry()));
+            }
+            result.complete(snapshotMetadataList);
+        } catch (IOException e) {
+            result.completeExceptionally(e);
+        }
+        return result;
+    }
+
+    @NotNull
+    private CompletableFuture<LedgerHandle> createLedger() {
+        CompletableFuture<LedgerHandle> future = new CompletableFuture<>();
+        bookKeeper.asyncCreateLedger(
+                config.getManagedLedgerDefaultEnsembleSize(),
+                config.getManagedLedgerDefaultWriteQuorum(),
+                config.getManagedLedgerDefaultAckQuorum(),
+                BookKeeper.DigestType.fromApiDigestType(config.getManagedLedgerDigestType()),
+                LedgerPassword,
+                (rc, handle, ctx) -> {
+                    if (rc != BKException.Code.OK) {
+                        future.completeExceptionally(bkException("Failed to create ledger", rc, -1));
+                    } else {
+                        future.complete(handle);
+                    }
+                }, null, null);
```

Comment:
Please add ledger metadata so that we can know which component is using this ledger through bookkeeper shell tools. 
We have different components using ledgers, we need to add metadata on the ledger to know who is using it. You can check `LedgerMetadataUtils` to learn about how to add it.

Reply:
I add metadata using `bucketKey`